### PR TITLE
btdu: init at 0.3.1

### PR DESCRIPTION
**Pull request by AtilaSaraiva**
###### Motivation for this change
I use the BTRFS filesystem and felt the need to have the tool [btdu](https://github.com/CyberShadow/btdu) installed, so I wrote this derivation. It is a very simple application that is very similar to [ncdu](https://dev.yorhel.nl/ncdu), but tailored for the BTRFS filesystem.
I hope it can be added to the repo to help more people.

###### Things done

Added the package btdu to the repository.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by symphorien — CHANGES_REQUESTED** [review]
Thanks for your PR. I left a few comments, but overall it looks good.

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:43]: Thx for the input!
It actually works without the ncurses as a buildInput, so that is why I kept it that way. 

I will test multiple variations of the d compilers to see what works.

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:33]: Thx for reviewing! Hm, I didn't think about that possibility. I will do that.

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [reply pkgs/tools/misc/btdu/default.nix:43]: buildInputs and nativeBuildInputs are the same except in some situations like cross-compilation for backward compatibility. To see the real behavior, set `strictDeps = true`.

**Comment by AtilaSaraiva** [comment]
I've reduced the number of dependencies as requested, but I'm still trying to figure out how to make an update script that works on multiple source files. Could u point me towards the right direction?

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:43]: that's still 2 compilers: gdc and dmd. According to the d wiki ldc is better than dmd, and gdc is too old to compiler btdu.

https://wiki.dlang.org/Compilers

**Comment by symphorien** [comment]
https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/plex/raw.nix#L59 random example of update script with multiple hashes

**Comment by symphorien** [comment]
Also the build output depends on the compiler (you can check with `nix-store --query --references ./result`) so could you remove them with `remove-references-to` ? (random example https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/radio/gnuradio/shared.nix#L87 )

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:43]: I actually tested removing one of them and it didn't work.

Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:43]: yeah I got to this point already haha, thx anyway! I'm now trying to figure out the reference thing u suggested I do. There isn't any sort of documentation on the script so it's been hard.

**Review by 06kellyjac — CHANGES_REQUESTED** [review]
Hello and Welcome.

Some notes:

pr title and commit should be `btdu: init at 0.3.1` with a separate commit for adding yourself to the maintainers list

for details please review the CONTRIBUTING doc

**Comment by AtilaSaraiva** [comment]
How should I proceed then? Remove the addition to the maintainers list and create another PRs with my addition?

**Comment by 06kellyjac** [comment]
no your addition to the maintainers list can stay in this PR, it just needs to be in a separate commit to the package addition

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:73]:
```suggestion
      platforms = platforms.linux;
```
as btrfs is linux specific

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:65]:
```suggestion
      ${removeReferencesTo}/bin/remove-references-to -t ${ldc} $(readlink -f $out/lib/libdruntime-ldc-shared)
```
(to be tested)

**Comment by symphorien** [comment]
Your PR should only contain two commits: the commit where you add yourself in the maintainer list, and the commit where you add btdu (with you as maintainer).

But now that you merged master into this branch it's hard.
```commands
# start from clean nixos-unstable
$ git checkout nixos-unstable
# create a new branch
$ git checkout -b foo
# readd yourself in the maintainer list
# commit
$ git add ...; git commit
# reapply the content of this PR
$ curl https://patch-diff.githubusercontent.com/raw/NixOS/nixpkgs/pull/157197.diff | patch -p1
$ git add ....; git commit -m "btdu: init at 0.3.1"
# now fix your current branch
$ git branch -f btdu HEAD
$ git checkout btdu
$ git push fork --force-with-lease
```

For subsequent updates to the "btdu: init at 0.3.1" please use `git commit --amend` and then `git push fork --force-with-lease`.

All this is not just annoying formal requirements, it helps reverting problematic commits in case of need.

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:73]: Oh yeah u right thx!

**Comment by AtilaSaraiva** [comment]
From the [06kellyjac](https://github.com/06kellyjac) comment, I should create another PR with the addition to the maintainers list, which is [here](https://github.com/NixOS/nixpkgs/pull/158278). I was expecting to be added to the maintainers list and update my branch.
Either way, should I close the other PR with the maintainer list addition or should close it and do it the way you explained?

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:65]: I will test right away!

Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:65]: I've tested here and it still references ldc:
```
/nix/store/jcb7fny2k03pfbdqk1hcnh12bxgax6vf-glibc-2.33-108
/nix/store/3y4g885qcv7wmzczcgiz4xgshy75snd3-ncurses-6.3
/nix/store/92wmg8niqrvwxsfigh5hhjpb0y9ska1i-zlib-1.2.11
/nix/store/dphabgk0ha7pzp4ca9sdry6fhvpin1lv-curl-7.81.0
/nix/store/z85pkakmhsj4pv8n44y7pw55hqvqn4zy-tzdata-2021e
/nix/store/4i6ff0409wip24r1hcjmych3mylw6i9s-ldc-1.27.1
```
this $out/lib/libdruntime-ldc-shared that actually confuses me the most.

Comment (symphorien) [reply pkgs/tools/misc/btdu/default.nix:65]: the second argument is the file that contains the reference. Either there is only one such file (find it with `nix why-depends`) or you can make a loop with find: random example found with grep:
`find $out -type f -exec ${removeReferencesTo}/bin/remove-references-to -t ${go} '{}' +`

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:62]:
```suggestion
    installPhase = ''
      runHook preInstall
      mkdir -p $out/bin
      cp btdu $out/bin/
      runHook postInstall
    '';
```otherwise postInstall is not run

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:57]: please add runHook for preBuild and postBuild

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [inline pkgs/tools/misc/btdu/default.nix:52]: please add runHook for preConfigure and postConfigure

**Comment by AtilaSaraiva** [comment]
Done! Sorry for the misundestanding, I'm still new to using git for big projects such as this one. The remove reference script still needs some work as it seems and I'm still working in the update script, I ask for a bit of patience.

**Comment by 06kellyjac** [comment]
There's no rush :)

**Comment by symphorien** [comment]
Yes please take the time you need, and thanks for bearing with us :)

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:57]: So u suggest having the configurePhase being a preBuild phase?

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:62]: Oh I didn't know that haha. Although I don't see anything to put on the preInstall hook, except in the case u want the buildPhase to be a preInstall hook. I will add the postInstall hook nevertheless.

Should I add the runHook even though the preInstall phase was not described?

**Comment by AtilaSaraiva** [comment]
No prob! I have another derivation already for publishing, I just taking this opportunity to learn and make a bunch of PRs correctly later.

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [reply pkgs/tools/misc/btdu/default.nix:57]: the default buildPhase contains `runHook preBuild` and `runHook postBuild`. If you overwrite the default buildPhase (which is perfectly fine) you must call the hooks manually (even if you don't use them, so that you can do `btdu.overrideAttrs(o: {preBuild = "foo"; })`).

File: pkgs/tools/misc/btdu/default.nix
Comment (symphorien) [reply pkgs/tools/misc/btdu/default.nix:62]: yes it's for downstream users adding their own preInstall (with .overrideAttrs)

File: pkgs/tools/misc/btdu/default.nix
Comment (AtilaSaraiva) [reply pkgs/tools/misc/btdu/default.nix:57]: Oh I see, Imma add them then.

**Comment by AtilaSaraiva** [comment]
I managed to finally remove the link to ldc and added all the hooks u can imagine and the program is working. Now I gonna focus on writing the update script. I wanted to simply grab the version and hash from the PKGBUILD on arch since it is maintained by the dev behind btdu, but I wonder if this is acceptable haha. Otherwise I might have to grab everything from github.

**Comment by symphorien** [comment]
> I wanted to simply grab the version and hash from the PKGBUILD on arch since it is maintained by the dev behind btdu, which would ensure it would build, but I wonder if this is acceptable haha.

Update scripts are not consumed by end users, only by you, the maintainer, so I don't see any reason to be picky.

**Review by symphorien — APPROVED** [review]
Thank you for your PR. It looks really good now.

I can merge right away (and let you add an update script on your next update PR) or wait for an update script now.

**Comment by AtilaSaraiva** [comment]
I still don't have an working update script :/. Can it be done after the merge while still in this PR or will I have to open another one?

**Comment by symphorien** [comment]
I would be another one.

**Comment by AtilaSaraiva** [comment]
Then wait a little bit, I think I can have this figured out until saturday.

**Comment by AtilaSaraiva** [comment]
Managed to write an update script, using python rather than shell (hope it is alright) since I'm a complete noob when it comes to shell. Seems like this is a somewhat novel approach!

**Comment by symphorien** [comment]
thanks!

**Pull request merged**
